File: leetcode/191.位1的个数/191.位1的个数.go
package main

func hammingWeight(num uint32) int {
	res := 0
	for num != 0 {
		num = num & (num - 1)
		res += 1
	}
	return res
}
func main() {
	n := 00000000000000000000000000001011
	result := hammingWeight(uint32(n))
	println(result)
}

File: leetcode/191.位1的个数/191.位1的个数.py
class Solution:
    """
    n & (n-1) 消除n的二进制表示中的最后一个1
    每消除一个1，res+=1
    """

    def hammingWeight(self, n: int) -> int:
        res = 0
        while n:
            n = n & (n - 1)
            res += 1
        return res


if __name__ == '__main__':
    n = 0o0000000000000000000000000001011
    result = Solution().hammingWeight(n)
    print(result)
